pkg/chains/formats/slsa/extract/extract.go

```diff
@@ -34,12 +34,67 @@ import (
 
 // SubjectDigests returns software artifacts produced from the TaskRun/PipelineRun object
 // in the form of standard subject field of intoto statement.
-// The type hinting fields expected in results help identify the generated software artifacts.
+// The type hinting fields expected in TaskRun results help identify the generated software artifacts in a TaskRun/PipelineRun.
 // Valid type hinting fields must:
 //   - have suffix `IMAGE_URL` & `IMAGE_DIGEST` or `ARTIFACT_URI` & `ARTIFACT_DIGEST` pair.
 //   - the `*_DIGEST` field must be in the format of "<algorithm>:<actual-sha>" where the algorithm must be "sha256" and actual sha must be valid per https://github.com/opencontainers/image-spec/blob/main/descriptor.md#sha-256.
 //   - the `*_URL` or `*_URI` fields cannot be empty.
 func SubjectDigests(ctx context.Context, obj objects.TektonObject) []intoto.Subject {
+	var subjects []intoto.Subject
+
+	switch obj.GetObject().(type) {
+	case *v1beta1.PipelineRun:
+		subjects = subjectsFromPipelineRun(ctx, obj)
+	case *v1beta1.TaskRun:
+		subjects = subjectsFromTaskrun(ctx, obj)
+	}
+
+	sort.Slice(subjects, func(i, j int) bool {
+		return subjects[i].Name <= subjects[j].Name
+	})
+
+	return subjects
+}
+
+func subjectsFromPipelineRun(ctx context.Context, obj objects.TektonObject) []intoto.Subject {
```

Comment:
The PipelineRun results should be taken into account as well.

Reply:
Do we want to support both pipelineRun results and taskrun results by default, OR do we want to make it a configmap field to let users select how Chains should work?

The latter sounds a bit better to me b/c I feel like the former might be a bit messy if users pull in catalog tasks that produce type hinting results, and users also define some overlapping type hinting results on the pipeline level.

Reply:
There are benefits in using the results from the PipelineRun. For example, one can use a Task that doesn't produce the expected Chains results (many of the Tasks in the tekton catalog don't).

We could make it configurable as long as the default includes the PipelineRun results so we don't break users. I don't really think making it configurable is necessary, but I don't have a strong opinion on this. More configuration options could make things more complicated for users though.

Reply:
Thanks @lcarva for sharing the thoughts!

> For example, one can use a Task that doesn't produce the expected Chains results (many of the Tasks in the tekton catalog don't).

That's a good point. I have thought about same thing in some catalog tasks. For example, current [git clone task](https://github.com/tektoncd/catalog/tree/main/task/git-clone/0.9#results) do output the repo uri and commit sha results, but it's not in type hinting format, whereas [kaniko task](https://github.com/tektoncd/catalog/tree/main/task/kaniko/0.6#results) fully supports this. 

In those cases, I think we should make sure at least some catalog tasks produce type hinting results for consistency and better user experience, instead of letting users to figure out what to do if some catalog tasks do not produce type hinting.

That said, I am not opposing the idea of keeping support of pipeline results and I am happy to keep supporting it. But at least for some catalog tasks, we should be consistent and make sure they produce results in type hinting way. 



> We could make it configurable as long as the default includes the PipelineRun results so we don't break users. I don't really think making it configurable is necessary, but I don't have a strong opinion on this. More configuration options could make things more complicated for users though.

Agreed. ConfigMap option seems to expose too much implementation details to users. But the risk is the messy state if type hinting exist on both taskrun and pipelinerun level. (i.e. author surfaces taskruns' type hinting results to pipelines' type hinting results).

Open to other thoughts and ideas here.

Reply:
Update from today's meeting: Let's make it configurable for now with the goal of making it not configurable in the future. This will allow for a safer feature rollout and give a chance for users to adapt. Since it's temporary, it should not increase maintenance long-term.

Reply:
@lcarva Done. Please take a look!